{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BC205: Algorithms for Bioinformatics.\n",
    "## IV. Motif Discovery\n",
    "\n",
    "### Christoforos Nikolaou\n",
    "\n",
    "#### Up to now we have seen how:\n",
    "* We define a sequence motif\n",
    "* We can search for known short motifs with a determined degree\n",
    "of ambiguity\n",
    "* We can estimate the existence of a motif in a sequence\n",
    "* We can define the strength of the motif in the sequence in an\n",
    "Entropy-based score\n",
    "\n",
    "![Motifs on Genome Sequences](figures/GenomicMotifs.PNG)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Consensus Sequences\n",
    "\n",
    "A _consensus sequence_ is a grammatically encoded pattern that represents the ambiguity of the pattern's instances in a given corpus. We usually encode consensus sequences in the form of regular expression notation. \n",
    "\n",
    "![Consensus Sequence Definition](figures/ConsensusSequence.png)\n",
    "\n",
    "#### Exercise\n",
    "\n",
    "1. Write the python code to:  \n",
    "   a. Read a set of aligned sequences of the same length  \n",
    "   b. Produce the consensus sequence   \n",
    "   c. Print it on the screen using regular expression notation  \n",
    "   d. Accompany the consensus sequence with a sequence bearing the most common residue in each position.  \n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "#### Creating of a PWM from a given sequnce\n",
    "\n",
    "Remember that a PWM is _MxN_ table, where _M_ is the number of the different residue types (e.g. _M=4_ for DNA) and _N_ is the length of the motif. Each position _(i,j)_ in a PWM holds the frequency of occurrence (probability) of residue _i_ to be found in position _j_.\n",
    "\n",
    "Such a table may be seen below \n",
    "![PWN](figures/PWM.PNG)\n",
    "\n",
    "Below we start by creating a function to read multiple fasta sequences in a list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# A function to read multiple fasta files each and return the sequences as a list\n",
    "def readmultifasta(file):\n",
    "    import regex as re\n",
    "    f = open(file, 'r')\n",
    "    seqs = []\n",
    "    for line in f:\n",
    "        x=re.match(\">\", line)\n",
    "        if x == None:\n",
    "            seqs.append(line.rstrip())\n",
    "    return(seqs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TTATAG', 'AGATAT', 'AGATAG', 'ATATCT', 'AGATAG', 'AGATAG', 'AGATAG', 'TGATAA', 'AGATAA', 'AGATAA', 'CGATAG', 'AGAGTT', 'TGATAA', 'TGATAA', 'AGATGG', 'AGATAG', 'AGATTG', 'AGATAA', 'TGATAA', 'AGATAA', 'AGATAG', 'TGATAG', 'TGATCA', 'TTATCA', 'AGATGG', 'TGATAT', 'AGATAG', 'TGATAA', 'GGATAC', 'AGATAA', 'CGATAA', 'TGATAG', 'AGATAA', 'TGATTA', 'AGATAA', 'AGATAG', 'TGATAT', 'AGATAA', 'TCAGAG', 'AAGTAG', 'AGATTA', 'TGATAG', 'TGATAG', 'AGATAC', 'TGATTG', 'AGATTA', 'AGAATA', 'AGATAA', 'AGATTA', 'AGCTTC']\n"
     ]
    }
   ],
   "source": [
    "seqs = readmultifasta(\"files/gata.fa\")\n",
    "print(seqs)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We next proceed with a function that takes a set of sequences of **the same length** and produces a PWM table from this set. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# A function to create a PWM from a set of aligned sequences of the same length\n",
    "def pwm(sequences):\n",
    "    nuc = ['A', 'C', 'G', 'T']\n",
    "    profile=[[0 for i in range(len(sequences[0]))] for j in range(len(nuc))]\n",
    "    #\n",
    "    for instance in sequences:\n",
    "        for j in range(len(instance)):\n",
    "            residue=instance[j]\n",
    "            profile[nuc.index(residue)][j]+=1\n",
    "            profile[nuc.index(residue)][j]=float(profile[nuc.index(residue)][j])\n",
    "    import numpy as np\n",
    "    pwm = np.array(profile)\n",
    "    pwm = pwm/len(sequences)\n",
    "    return(pwm)\n",
    "\n",
    "mypwm=pwm(seqs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[0.6  0.02 0.96 0.02 0.72 0.44]\n",
      " [0.04 0.02 0.02 0.   0.06 0.06]\n",
      " [0.02 0.9  0.02 0.04 0.04 0.4 ]\n",
      " [0.34 0.06 0.   0.94 0.18 0.1 ]]\n"
     ]
    }
   ],
   "source": [
    "print(mypwm)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice how we assign nucleotides at the beginning alphabetically ordered, which means that \"A\", corresponds to the first row, \"C\" to the second etc."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Transformation of PWM into PSSM on the basis of a given sequence\n",
    "\n",
    "We can transform this table by applying a normalization and log-transformation against nucleotide occurrences from a given sequence, to create a PSSM matrix.\n",
    "\n",
    "PSSM matrices are constructs that are context-dependent, meaning that they can better capture a motif match in a given sequence, provided they have incorporated that sequence's nucleotide background composition.\n",
    "\n",
    "![PSSM Matrix](figures/PSSM.PNG)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "We shall first write a function to read the target sequence in fasta format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def readfasta(fastafile):\n",
    "    import regex as re\n",
    "    f=open(fastafile, 'r')\n",
    "    seq = \"\"\n",
    "    total = 0\n",
    "    for line in f:\n",
    "        x=re.match(\">\", line)\n",
    "        if x == None:\n",
    "            length=len(line)\n",
    "            total=total+length\n",
    "            seq=seq+line[0:length-1]\n",
    "    seq=seq.replace('N','')\n",
    "    f.close()\n",
    "    return(seq)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And then write another function to calculate nucleotide frequencies on a given input sequence/genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculating mononucleotid composition\n",
    "def nuccomp(sequence):\n",
    "    import numpy as np\n",
    "    nucfreq = [0, 0, 0, 0]\n",
    "    nuc = ['A', 'C', 'G', 'T']\n",
    "    for i in range(len(nuc)):\n",
    "        nucfreq[i]=sequence.count(nuc[i])\n",
    "    nucfreq=np.array(nucfreq)/len(sequence)\n",
    "    return(nucfreq)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can then run them sequentially"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.33585935 0.16534241 0.16384496 0.33495328]\n"
     ]
    }
   ],
   "source": [
    "targetsequence=readfasta(\"files/staaur.fa\")\n",
    "nucfreqs=nuccomp(targetsequence)\n",
    "print(nucfreqs)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alternatively we can deploy the kmers function from a previous class and run it for k=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def kmers(genomefile, k):\n",
    "    import regex as re\n",
    "\n",
    "    file = open(genomefile, 'r')\n",
    "\n",
    "    seq = \"\"\n",
    "    kmertable = {} \n",
    "\n",
    "    count = 0\n",
    "    for line in file:\n",
    "        count +=1\n",
    "        if (count > 1) :\n",
    "            length=len(line)\n",
    "            seq=seq+line[0:length-1]\n",
    "            \n",
    "    file.close()\n",
    "\n",
    "    seq = re.sub(\"[^AGCT]\", \"\", seq)\n",
    "\n",
    "    for i in range(len(seq)-k):\n",
    "        DNA=seq[i:i+k]\n",
    "        if DNA not in kmertable.keys():\n",
    "            kmertable[DNA]=1\n",
    "        else:\n",
    "            kmertable[DNA]+=1\n",
    "\n",
    "    kmertable = {k: float(v) / len(seq) for k, v in kmertable.items()}\n",
    "    kmertable = dict(sorted(kmertable.items()))\n",
    "    return(list(kmertable.values()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nucfreqs = kmers(\"files/Staaur.fa\", 1)\n",
    "print(nucfreqs)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Creation of a PSSM\n",
    "\n",
    "Î¤he next obvious step is to combine the PWM with the Array of the nucleotide composition of the target sequnce and log-transform the resulting table into a PSSM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creation of a PSSM\n",
    "def pssm(pwm, nucfreqs):\n",
    "    import numpy as np\n",
    "    import math\n",
    "    pseudocount=0.01\n",
    "    pssm=[[0 for i in range(len(pwm[0]))] for j in range(len(nucfreqs))]\n",
    "    for i in range(len(nucfreqs)):\n",
    "        pssm[i]=(np.array(pwm[i])+pseudocount)/nucfreqs[i]\n",
    "    for i in range(len(pssm)):\n",
    "        for k in range(len(pssm[0])):\n",
    "            pssm[i][k]=math.log(pssm[i][k])/math.log(2)\n",
    "    return(np.array(pssm))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 1.31059538 -3.03517946  1.97977088 -3.03517946  1.5696826   0.87171114]\n",
      " [-2.34326171 -3.0802273  -3.0802273  -4.66518981 -1.85783488 -1.85783488]\n",
      " [-3.0801393   1.84269284 -3.0801393  -2.3431737  -2.3431737   0.69245021]\n",
      " [ 0.50493453 -1.81699356 -4.62434848  1.94550713 -0.37642097 -1.16491686]]\n",
      "[[ 0.86095362 -3.48482122  1.53012912 -3.48482122  1.12004084  0.42206938]\n",
      " [-1.72545683 -2.46242243 -2.46242243 -4.04738493 -1.24003001 -1.24003001]\n",
      " [-2.4492969   2.47353524 -2.4492969  -1.7123313  -1.7123313   1.32329261]\n",
      " [ 0.06339504 -2.25853305 -5.06588798  1.50396763 -0.81796046 -1.60645636]]\n"
     ]
    }
   ],
   "source": [
    "nucfreqs0 = kmers(\"files/Ecoli.fa\", 1)\n",
    "nucfreqs1 = kmers(\"files/Staaur.fa\", 1)\n",
    "\n",
    "mypssm=pssm(mypwm, nucfreqs0)\n",
    "print(mypssm)\n",
    "\n",
    "mypssm=pssm(mypwm, nucfreqs1)\n",
    "print(mypssm)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Searching a sequence with a PWM/PSSM \n",
    "\n",
    "In the following, we define a pssmSearch function that takes as input a PSSM, a target sequence and a score threshold which we set to be a percentage of the maximum possible PSSM score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Update to allow for percentage thresholds\n",
    "\n",
    "def pssmSearch(pssm, sequence, threshold):\n",
    "\n",
    "    nuc = ['A', 'C', 'G', 'T']\n",
    "    hits = []\n",
    "    instances = []\n",
    "    x = []\n",
    "    allscores = [] # for plotting reasons\n",
    "    \n",
    "    # Step 1: Calculation of maximum possible PSSM score\n",
    "    maxPssm = 0\n",
    "    for j in range(len(pssm[0])):\n",
    "        maxPssm = maxPssm + max(pssm[:,j])\n",
    "    \n",
    "    # Step 2: Search\n",
    "    for i in range(len(sequence)-len(pssm[0])):\n",
    "        x.append(i)\n",
    "        instance=sequence[i:i+len(pssm[0])]\n",
    "        score=0\n",
    "        for l in range(len(instance)):\n",
    "            score=score+pssm[nuc.index(instance[l])][l]\n",
    "        if (score > threshold*maxPssm):\n",
    "            hits.append(i)\n",
    "            instances.append(instance)\n",
    "            allscores.append(score)\n",
    "    \n",
    "    return(hits, instances, allscores)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAG',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAG',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAG',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAG',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'AGATAG',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAG',\n",
       " 'AGATAG',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAG',\n",
       " 'TGATAG',\n",
       " 'TGATAG',\n",
       " 'AGATAG',\n",
       " 'TGATAG',\n",
       " 'TGATAG',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAG',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAG',\n",
       " 'AGATAG',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'AGATAG',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAG',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'AGATAG',\n",
       " 'AGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAG',\n",
       " 'AGATAG',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAG',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAG',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " 'TGATAG',\n",
       " 'AGATAG',\n",
       " 'TGATAG',\n",
       " 'AGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'TGATAA',\n",
       " 'AGATAA',\n",
       " 'AGATAA',\n",
       " ...]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pssmSearch(mypssm, targetsequence, 0.80)[1]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exercise\n",
    "\n",
    "Perform the same process as above but now using:  \n",
    "    a. A fixed most frequent word pattern instead of a PSSM   \n",
    "    b. A Hamming Distance approach for matches   \n",
    "    c. As in this example, allow for different levels of Hamming Distance to report matches. "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Entropy calculations\n",
    "\n",
    "Shannon Entropy and Information Content are quantities that we use to assess the ambiguity in a motif, mostly in the form of a PWM matrix. \n",
    "\n",
    "Remember the definition of Shannon Entropy:\n",
    "\n",
    "![Shannon Entropy](figures/ShannonEntropyFormula.PNG)\n",
    "\n",
    "We can then apply Entropy and Information Content calculations on the resulting hits/matches.\n",
    "\n",
    "![Information](figures/InformationFormula.PNG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Entropy and Information Content of a PWM\n",
    "\n",
    "def pwmEntropyInformation(pwm):\n",
    "    \n",
    "    import numpy as np\n",
    "    k = pwm.shape[1]\n",
    "\n",
    "    information = np.zeros([1,k]) #computing the information of each position\n",
    "    for i in range(k):\n",
    "        information[0,i] = 2-abs(sum([elem*np.log2(elem) for elem in pwm[:,i] if elem > 0]))\n",
    "    \n",
    "    sumInfo = np.sum(information)\n",
    "    scaledSumInfo = sumInfo/k\n",
    "    \n",
    "    return(information, sumInfo, scaledSumInfo)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### PWM vs PSSM Searches\n",
    "\n",
    "Lets compare the output of searches with PWM and PSSM. Our previously defined function can handle both. Below we search for instances of the motifs from a starting PWM, PSSM at the top 90% similarity threshold. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "pwmsearches = pssmSearch(mypwm, targetsequence, 0.90)[1]\n",
    "pssmsearches = pssmSearch(mypssm, targetsequence, 0.90)[1]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below we print our output in a file to use in external sources"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(pwmsearches)\n",
    "f = open(\"PWMSearches.txt\", \"w\")\n",
    "for instance in pwmsearches:\n",
    "    \tf.write(instance+\"\\n\")\n",
    "f.close()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And now we check the Information Content of each of the two search results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([[1.0231508 , 2.        , 2.        , 2.        , 2.        ,\n",
       "         0.28962496]]),\n",
       " 9.312775760245215,\n",
       " 1.5521292933742024)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwmEntropyInformation(pwm(pwmsearches))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([[1.00319023, 2.        , 2.        , 2.        , 2.        ,\n",
       "         1.37981562]]),\n",
       " 10.383005849751157,\n",
       " 1.7305009749585263)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwmEntropyInformation(pwm(pssmsearches))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From which we see that the PSSM search at the same threshold returns a more restricted and informative set of motif instances."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The next problem. Discover a new motif from a given set of sequences\n",
    "\n",
    "#### Part 1. Formulating the problem\n",
    "1. Given a set of sequences that each contains an instance of the motif, find the motif.\n",
    "\n",
    "#### A Naive Brute-Force Approach\n",
    "\n",
    "Find the most common k-mer in a set of sequences by iteratively scanning all the sequences\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Given a set of s sequences: Find a set of k-mers (for a given\n",
    "length k, one from each sequence) that maximizes the score (or\n",
    "minimizes the distance) of each (one) k-mer with its sequence\n",
    "2. Collect k-mers\n",
    "3. Create a motif from them\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Brute Force Approach\n",
    "\n",
    "What is the complexity of the BFA?\n",
    "1. Number of k-mers 4k\n",
    "2. Number of k-mers in each sequence: (n â k + 1)\n",
    "3. Number of calculations for each k-mer given s sequences of\n",
    "length n: (n â k + 1)s\n",
    "4. Total number of calculations 4k (n â k + 1)s\n",
    "\n",
    "The complexity of the algorithm is at least O(ns).\n",
    "\n",
    "We need something faster!\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### A more nuanced Brute-Force Approach : Median String\n",
    "\n",
    "Find a set of k-mers, one from each sequence, that together, create a PWM that has maximal Information content (or, alternatively, minimum distance).\n",
    "\n",
    "A string that **minimizes the sum of distances to the strings of a given set** is known as (generalized) median string of the set. This is proven to be NP-hard and so no efficient algorithms exist.\n",
    "\n",
    "* Assuming we have a way to calculate the distance of a k-mer k\n",
    "\n",
    "```\n",
    "from a given sequence seq  \n",
    "    for k in kmers:  \n",
    "        for seq in sequences:  \n",
    "            if distance(k, seq)<min_distance:  \n",
    "                min_distance<-distance(k,seq)  \n",
    "                motif[seq]<-k\n",
    "```\n",
    "\n",
    "* Because each k-mer needs to pass only once through each\n",
    "sequence, the median string has O(4k) complexity because k is\n",
    "(usually) much shorter than the length of the sequence.\n",
    "\n",
    "* However, it is still quite slow and for k>10 its implementation\n",
    "is still unapplicable.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# code"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### An alternative\n",
    "\n",
    "* Assume a greedy approach to go through all sequences\n",
    "updating a motif every time\n",
    "\n",
    "* Starting from sequence i:\n",
    "\n",
    "1. find the most common k-mer\n",
    "2. create a profile from it (adding pseudocounts to all 0-values)\n",
    "3. go to the next sequence\n",
    "4. choose the k-mer that best fits the profile\n",
    "5. store that k-mer in the collection and update profile\n",
    "6. iterate steps 3->5.\n",
    "\n",
    "* Weâ ve just described a Greedy approach for discovering a\n",
    "motif p of a given length k among t sequences.\n",
    "\n",
    "* Which problems you see in this approach?\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Greedy Approach. Problems\n",
    "\n",
    "* Dependence on the starting sequence. For robustness we can repeat the process above with a different starting sequence every time.\n",
    "* It expects most (if not all) sequences in the collection to contain the motif. This doesn't happen often.\n",
    "* It will work worse as the motif becomes \"degenerate\" of low information content\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Analysis. Greedy Approach\n",
    "\n",
    "1. Why Greedy: It takes kmers from the first sequence only to\n",
    "scan in the following. Thus it doesnât go through all\n",
    "combinations of sequences and k-mers. As weâve seen above\n",
    "the trade-off is speed.\n",
    "2. KEY: It assumes that all sequences contain the motif. If the\n",
    "first sequence doesnât contain the motif (in any variation) then\n",
    "we are doomed in looking for something that is non-sensical.\n",
    "3. A way to go around this is to sample a small percentage of\n",
    "sequences randomly, which brings us to the next-to-last chapter\n",
    "of the motif finding problem\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### A randomized approach\n",
    "\n",
    "* In the Greedy Approach we take the kmers from the first\n",
    "sequence and scan over the rest. In this way an initial wrong\n",
    "choice may lead you to disastrous results.\n",
    "* In a Randomized Approach we start, instead with a\n",
    "collection of s k-mers, one from each sequence, build a profile,\n",
    "scan the sequences with that profile, update it and repeat until\n",
    "the k-mer set is good enough match for the updated profile.\n",
    "* Stop and think of *the problems we get rid of* with this\n",
    "approach.\n",
    "\n",
    "![Gibbs Sampling for Motif Identification](figures/GibbsSampler.PNG)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Pseudocode\n",
    "\n",
    "```\n",
    "for seq in sequences:  \n",
    "    profile[seq]<-random(k, seq)  \n",
    "    while distance(profile, sequences)>threshold  \n",
    "        for seq in sequences:  \n",
    "        profile[seq]<-max(k, profile, seq)  \n",
    "```"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* You can think of alternatives to \"distance\" with Information Content.\n",
    "* The main idea is unchanged. \n",
    "  1. You start with a random collection of k-mers, one from each sequence\n",
    "  2. You create a PWM out of these\n",
    "  3. You scan each sequence with the PWM and substitute the chosen k-mer with the one that best fits the PWM\n",
    "  4. You update the PWM and calculate its Information Content\n",
    "  5. If I >= threshold stop. Otherwise repeat 3-5.  "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "orig_nbformat": 2,
  "vscode": {
   "interpreter": {
    "hash": "fb833273add3e7c60eb33c0608260b79a61e072ade6f02cc8d07b0a26eef8ab8"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
