{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Code for exercises, small projects or variations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dichotomous search with indexing (Chapter 1) [Correct, as of 8.3.2023]\n",
    "\n",
    "## Part 1. Indexing ecoli genome\n",
    "\n",
    "### reading of file\n",
    "file = open('files/ecoli.fa', 'r')\n",
    "ecoli = ''\n",
    "count = 0\n",
    "for line in file:\n",
    "    count += 1\n",
    "    if (count > 1): # the first line contains the non-sequence header so we discard it \n",
    "        ecoli += line.replace(\"\\n\", \"\") # we string the newline character from the end of each line\n",
    "\n",
    "\n",
    "### Creating a sort list of all k-mers in the genome \n",
    "k = 10\n",
    "# note that position i is attached to kmer\n",
    "kmers = [ecoli[i:i+k]+\":\"+str(i) for i in range(len(ecoli)-k+1)] \n",
    "\n",
    "dictkmers = {}\n",
    "\n",
    "for k in kmers:\n",
    "    kstring = k.split(\":\")[0]\n",
    "    dictkmers[kstring] = \"\"\n",
    "\n",
    "### Recasting the kmer list to a dictionary with kmers as keys and position list as values\n",
    "for k in kmers:\n",
    "    kstring = k.split(\":\")[0]\n",
    "    dictkmers[kstring] = dictkmers[kstring]+\",\"+str(k.split(\":\")[1])\n",
    "\n",
    "list_of_kmers = list(dictkmers.keys()) # full list\n",
    "list_of_kmers.sort()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Part 2. Dichotomous Search for k-mers [Correct, as of 8.3.2023]\n",
    "\n",
    "### Using time to measure time of execution\n",
    "import time\n",
    "start_time = time.time()\n",
    "\n",
    "### Pattern search\n",
    "\n",
    "pattern = 'AAAAAACGAG' # non-mer\n",
    "pattern = 'ATGCACGACC' # existing\n",
    "\n",
    "matches = 0\n",
    "\n",
    "iter = 0\n",
    "min = 1\n",
    "max = len(list_of_kmers)\n",
    "\n",
    "midpoint = int((max+min)/2)\n",
    "\n",
    "import math \n",
    "\n",
    "while iter <= math.log2(len(list_of_kmers)) + 1:\n",
    "    kmer = list_of_kmers[midpoint]\n",
    "    iter += 1\n",
    "    if (pattern == kmer): # case is a success\n",
    "        matches = len(dictkmers[kmer][1:].split(\",\")) # number of matches equals length of positions\n",
    "        positions = dictkmers[kmer][1:].split(\",\")\n",
    "        print(\"Pattern\", pattern,\" matched \", matches, \" times in position(s):\", positions)\n",
    "        break\n",
    "    if (pattern > kmer):\n",
    "        min = midpoint\n",
    "        midpoint = int((max+min)/2)\n",
    "    if (pattern < kmer ):\n",
    "        max = midpoint\n",
    "        midpoint = int((max+min)/2)\n",
    "if (matches == 0):\n",
    "    if (pattern == list_of_kmers[0]):\n",
    "        kmer = list_of_kmers[0]\n",
    "        matches = len(dictkmers[kmer][1:].split(\",\")) # number of matches equals length of positions\n",
    "        positions = dictkmers[kmer][1:].split(\",\")         \n",
    "        print(\"Pattern\", pattern,\" matched \", matches, \" times in position(s):\", positions)\n",
    "    if (pattern == list_of_kmers[-1]):\n",
    "        kmer = list_of_kmers[-1]\n",
    "        matches = len(dictkmers[kmer][1:].split(\",\")) # number of matches equals length of positions\n",
    "        positions = dictkmers[kmer][1:].split(\",\")         \n",
    "        print(\"Pattern\", pattern,\" matched \", matches, \" times in position(s):\", positions)\n",
    "    else:\n",
    "        print(\"No matches found\")\n",
    "\n",
    "print(\"--- %s seconds ---\" % (time.time() - start_time))\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice how the above strategy of _dichotomous search_ is actually one order of magnitude faster than a regular expression search performed with a python generator function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pattern search with a generator function\n",
    "import re\n",
    "import regex\n",
    "\n",
    "import time\n",
    "start_time = time.time()\n",
    "\n",
    "keystring = 'ATGCACGACC'\n",
    "# generator function with re.finditer()\n",
    "dictkmers[keystring] = [m.start() for m in re.finditer(f'(?={keystring})', ecoli)]\n",
    "\n",
    "matches = len(dictkmers[keystring]) \n",
    "print(keystring, \" was found \", matches, \" times at positions:\", dictkmers[keystring])\n",
    "print(\"--- %s seconds ---\" % (time.time() - start_time))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Clustering of a dataset \n",
    "\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "from scipy.cluster.hierarchy import linkage, dendrogram\n",
    "\n",
    "# Load the dataframe and assign values/labels\n",
    "df = pd.read_csv('files/GCContent_simple.csv')\n",
    "dvalues = df['GCContent'].values.reshape(-1,1)\n",
    "dlabels = list(df['Genome'])\n",
    "\n",
    "# Calculate the distances\n",
    "distances = pdist(dvalues)\n",
    "\n",
    "# Convert the pairwise distances into a square distance matrix\n",
    "distance_matrix = squareform(distances)\n",
    "\n",
    "# Calculate the linkage matrix using Ward's method\n",
    "linkage_matrix = linkage(distance_matrix, method='ward')\n",
    "\n",
    "# Plot the dendrogram\n",
    "sns.set_style('white')\n",
    "dendrogram(linkage_matrix, labels=dlabels, color_threshold=0, orientation='left')\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PSSM Search\n",
    "\n",
    "# read file\n",
    "import sys\n",
    "import re\n",
    "import math\n",
    "import itertools\n",
    "import numpy as np\n",
    "\n",
    "#threshold=raw_input(\"Provide Score Threshold: \")\n",
    "threshold=float(sys.argv[3])\n",
    "\n",
    "pssm=np.genfromtxt(sys.argv[2], names=True, delimiter='\\t', dtype=None)\n",
    "size=len(pssm)\n",
    "maxscore=0\n",
    "\n",
    "for k in range(size):    \n",
    "    maxscore=maxscore+max(pssm[k][1],pssm[k][2],pssm[k][3],pssm[k][4])\n",
    "\n",
    "f = open(sys.argv[1],\"r\")\n",
    "seq = \"\"\n",
    "total = 0\n",
    "\n",
    "p=re.compile(\">(.+)\")\n",
    "\n",
    "for line in f:\n",
    "\tx=p.search(line)\n",
    "\tif x:\n",
    "\t\tname=x.group(1)\n",
    "\tif x == None:\n",
    "\t\tlength=len(line)\n",
    "\t\ttotal=total+length\n",
    "\t\tseq=seq+line[0:length]\n",
    "\t\tscore=[0 for x in range(len(seq)-size)];\n",
    "\n",
    "\t\tfor i in range(len(seq)-size):\n",
    "\t\t\tpattern=\"\"\n",
    "\t\t\tfor j in range(size):\n",
    "\t\t\t\tpattern=pattern+seq[i+j]\n",
    "\t\t\t\tscore[i]=score[i]+pssm[seq[i+j]][j]\n",
    "\t\t\t\n",
    "\t\t\tif score[i]/maxscore >= threshold:\n",
    "\t\t\t\tprint name,\"\\t\",i+1,\"\\t\",pattern,\"\\t\",score[i]/maxscore\n",
    "\t\t\n",
    "\t\tseq = \"\"\n",
    "\t\ttotal = 0\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### A divide-and-conquer with binary splits for the location of parity transitions (NOT WORKING)\n",
    "\n",
    "def dcParity(genomefile, type, window, resolution):\n",
    "    \n",
    "    import matplotlib.pyplot as plt\n",
    "    import regex as re  \n",
    "\n",
    "    # Reading of the sequence\n",
    "    f = open(genomefile, 'r')\n",
    "    seq = \"\"\n",
    "    total = 0\n",
    "    A=T=G=C=[]\n",
    "    times=0;\n",
    "    for line in f:\n",
    "        x=re.match(\">\", line)\n",
    "        if x == None:\n",
    "            length=len(line)\n",
    "            total=total+length\n",
    "            seq=seq+line[0:length-1]\n",
    "    f.close()\n",
    "\n",
    "    # Parity Calculations (for plotting purposes)\n",
    "    x = []\n",
    "    parity = []\n",
    "    step = int(window/10)\n",
    "    times = int(len(seq)/step);\n",
    "    for i in range(times):\n",
    "        x.append(i*step)\n",
    "        DNA=seq[i*step:i*step+window]\n",
    "        A=DNA.count(\"A\")\n",
    "        T=DNA.count(\"T\")\n",
    "        C=DNA.count(\"C\")\n",
    "        G=DNA.count(\"G\")\n",
    "        if type == \"AT\":\n",
    "            parity.append(float(A-T)/float(A+T))\n",
    "        if type == \"GC\":\n",
    "            parity.append(float(G-C)/float(G+C))\n",
    "\n",
    "    # Main part that performs the identification of transitions\n",
    "    transitions = []\n",
    "    grange = len(parity)\n",
    "    start = 0\n",
    "    end = len(parity)-1\n",
    "    midpoint  = int(len(parity)/2)\n",
    "    while (grange > resolution):\n",
    "        print(midpoint, grange, resolution)\n",
    "        if (parity[start]*parity[midpoint] < 0):\n",
    "            grange = midpoint - start \n",
    "            midpoint = int(midpoint/2)\n",
    "        if (parity[midpoint]*parity[end] < 0):\n",
    "            grange = end - midpoint\n",
    "            midpoint = end-int(midpoint/2)\n",
    "    \n",
    "    transitions.append(midpoint)\n",
    "        \n",
    "    # Plotting Parity and Predicted Transitions\n",
    "    plt.plot(x, parity, color= \"green\", linewidth = 3.0)\n",
    "    plt.axhline(y = 0, color = 'grey', linewidth = 2.0)\n",
    "\n",
    "    # plotting vertical lines at Ori and Ter\n",
    "    for site in transitions:\n",
    "        plt.axvline(x = site, color = 'black')\n",
    "\n",
    "    #plt.scatter(x, parity, color= \"green\")\n",
    "\n",
    "    # x-axis label\n",
    "    plt.xlabel('Genome Coordinates')\n",
    "    # frequency label\n",
    "    plt.ylabel('Relative Difference')\n",
    "    # plot title\n",
    "    plt.title('Parity')\n",
    "    # showing legend\n",
    "    #plt.legend()\n",
    "\n",
    "    # function to show the plot\n",
    "    plt.show()\n",
    "\n",
    "    return(transitions)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scrapbook\n",
    "#### Bits of code to use for help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(list_of_kmers[1:100])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(kmers[0:10])\n",
    "list_of_kmers.count('AAAAAAAAAC')\n",
    "\n",
    "print(ecoli[:1000])\n",
    "\n",
    "keystring = 'AAAAAAAAAC'\n",
    "matches = re.findall(keystring, str(ecoli))\n",
    "print(matches)\n",
    "\n",
    "positions = [m.start() for m in re.finditer(f'(?={keystring})', ecoli)]\n",
    "print(positions)\n",
    "\n",
    "#string='atha'\n",
    "#[m.start() for m in re.finditer(f'(?={string})', 'athathaathan')]\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gibbs Sampling to locate a motif in a set of sequences\n",
    "# code\n",
    "sequences = []\n",
    "with open ('files/motifs_in_sequence.fa') as file:\n",
    "    for line in file:\n",
    "        sequences.append(line.strip()) # 50 sequences as elements of a list. 100 bases each sequence\n",
    "    \n",
    "\n",
    "##Gibbs sampler##\n",
    "import random\n",
    "import numpy as np\n",
    "\n",
    "def Gibbs_sampler(sequences,k): #k is the length of the motif, sequence is a list with the sequences\n",
    "    \n",
    "    dictionary = {'A':0,\n",
    "                  'T':1,\n",
    "                  'C':2,\n",
    "                  'G':3}\n",
    "    \n",
    "    column_sum = len(sequences) #number of rows (50) or number of sequences\n",
    "    length = len(sequences[0]) #number of columns or number of nucleotides in seq\n",
    "    Imax = 1.8*k #threshold of I\n",
    "    \n",
    "    pwm = np.zeros([4,k]) # A,T,C,G X len(motif)\n",
    "    \n",
    "    for seq in sequences:\n",
    "        rand_start = random.randint(0, length-k) #pick a random nucleotide from each sequence\n",
    "        motif = seq[rand_start:rand_start+k] #and take substring as the motif\n",
    "       \n",
    "        lst = enumerate(motif) #finding the index of each nucleotide in the motif to access the correct column\n",
    "                               #and using the dictionary to access the correct row \n",
    "        for i in lst: #making the first random pwm\n",
    "            pwm[dictionary[i[1]],i[0]]+=1\n",
    "            \n",
    "    pwm = pwm/column_sum\n",
    "    \n",
    "    information = np.zeros([1,k])\n",
    "    count=0\n",
    "    while (np.sum(information)) < Imax: #while information_content of the pwm \n",
    "        motives=[]                      #is less than the threshold\n",
    "        \n",
    "        information_old = np.sum(information) #keeping the previous value of information contect\n",
    "                                             #to check convergence in case the theshold \n",
    "        for row in range (column_sum):                #is never reached\n",
    "            maxx=0\n",
    "            rand_seq = random.randint(0, column_sum-1) #pick a random index - sequence\n",
    "            seq = sequences[rand_seq]               \n",
    "            for i in range(len(seq)-k):   #take each k-mer from the sequence \n",
    "                score = 0\n",
    "                substring = seq[i:i+k]\n",
    "                lst = enumerate(substring)\n",
    "                \n",
    "                for j in lst:                         #scoring each k-mer based on the pwm\n",
    "                    score+=pwm[dictionary[j[1]],j[0]]   #keeping the motif with the highest score\n",
    "                                                         #from each sequence\n",
    "                if score > maxx:  \n",
    "                    maxx = score\n",
    "                    motif = substring\n",
    "                    \n",
    "            motives += [motif] #keep all the motifs with the highest score in the list motives\n",
    "        \n",
    "        pwm = np.zeros([4,k]) # A,T,C,G X len(motif) \n",
    "        \n",
    "        for elem in motives: \n",
    "            lst = enumerate(elem)\n",
    "            for i in lst:         #making the new pwm\n",
    "                pwm[dictionary[i[1]],i[0]]+=1\n",
    "                 \n",
    "        pwm = pwm/column_sum\n",
    "        \n",
    "        information = np.zeros([1,k]) #computing the information of each position\n",
    "        for i in range(k):\n",
    "            information[0,i] = 2-abs(sum([elem*np.log2(elem) for elem in pwm[:,i] if elem > 0]))\n",
    "            \n",
    "\n",
    "        if abs(information_old - np.sum(information)) <= 0.5: #ckecking convergence \n",
    "            count+=1\n",
    "            if count == 10: #if the difference of the information content is less or equal to 0.5\n",
    "                break       #for consecutive 10 iterations then break\n",
    "        else:\n",
    "            count=0\n",
    "    \n",
    "    max_index_col = np.argmax(pwm, axis=0) #extracting the motif according to the   \n",
    "                                           #highest frequency of each nucleotide in each position\n",
    "    motif=''\n",
    "    for values in max_index_col:\n",
    "        for keys in dictionary.keys():\n",
    "            if values == dictionary[keys]:\n",
    "                motif+= keys\n",
    "        \n",
    "    return pwm,information,motif\n",
    "\n",
    "\n",
    "#repeat the algorithmm 100 times for each k (3 to 7) and keep the pwm and motif with the highest infromation_content\n",
    "#this process takes approximately 4min (in my computer)\n",
    "    \n",
    "####100-cycled GIbbs#####\n",
    "for k in range (3,8):\n",
    "    highest_info = 0\n",
    "    for i in range (100):\n",
    "        summ=0\n",
    "        pwm, information_content,motif = Gibbs_sampler(sequences,k)\n",
    "        summ+=np.sum(information_content)\n",
    "        if summ > highest_info:\n",
    "            highest_info = summ\n",
    "            pwm_ret = pwm\n",
    "            motif_ret = motif\n",
    "        \n",
    "    print('\\nThe information content of the motif divided by it\\'s length is:',highest_info/k) #divide by length to normalize and compare among other k\n",
    "    print('The pwm of the motif is:\\n',pwm_ret)\n",
    "    print('The motif is:',motif_ret)\n",
    "\n",
    "\n",
    "\n",
    "#To find the motifs for each k with the highest information content i have to repeat gibbs sampler many times because of the randomness that takes place\n",
    "#The motif that returns the highest scaled information content is GAT (k = 3, I/k = 1.9528 or 2!) but i think that the motif that we are looking for\n",
    "#is the motif GATA (k=4, I/k = 1.857) which contains GAT and is contained in all the longer found motifs.\n",
    "#(Also we know the existence of the GATA transcription factors and indeed GATA is part of these binding sites)\n",
    "#Sometimes the algorithm returns other 3-mers (eg ATG,GGC,AAG) with high information content as well but it\n",
    "#is noted that GATA is returned almost in every repetition of the 100-cycled Gibbs which means that it may be the only 4-mer motif with so high information content.\n",
    "#Finally i have to report that the threshold that i chose for checking convergence may not be the best choice and should be stricter\n",
    "#Either way i kept this threshold because otherwise i would have made the algorithm a lot slower (is already slow though :P)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "==============================\n",
    "This is a dynamic programming algorithm for finding the optimal alignment of\n",
    "two strings.\n",
    "Example\n",
    "-------\n",
    "    >>> x = \"GATTACA\"\n",
    "    >>> y = \"GCATGCU\"\n",
    "    >>> print(nw(x, y))\n",
    "    G-ATTACA\n",
    "    GCA-TGCU\n",
    "\"\"\"\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "def nw(x, y, match = 1, mismatch = 1, gap = 1):\n",
    "    nx = len(x)\n",
    "    ny = len(y)\n",
    "    # Optimal score at each possible pair of characters.\n",
    "    F = np.zeros((nx + 1, ny + 1))\n",
    "    F[:,0] = np.linspace(0, -nx * gap, nx + 1)\n",
    "    F[0,:] = np.linspace(0, -ny * gap, ny + 1)\n",
    "    # Pointers to trace through an optimal aligment.\n",
    "    P = np.zeros((nx + 1, ny + 1))\n",
    "    P[:,0] = 3\n",
    "    P[0,:] = 4\n",
    "    # Temporary scores.\n",
    "    t = np.zeros(3)\n",
    "    for i in range(nx):\n",
    "        for j in range(ny):\n",
    "            if x[i] == y[j]:\n",
    "                t[0] = F[i,j] + match\n",
    "            else:\n",
    "                t[0] = F[i,j] - mismatch\n",
    "            t[1] = F[i,j+1] - gap\n",
    "            t[2] = F[i+1,j] - gap\n",
    "            tmax = np.max(t)\n",
    "            F[i+1,j+1] = tmax\n",
    "            if t[0] == tmax:\n",
    "                P[i+1,j+1] += 2\n",
    "            if t[1] == tmax:\n",
    "                P[i+1,j+1] += 3\n",
    "            if t[2] == tmax:\n",
    "                P[i+1,j+1] += 4\n",
    "    # Trace through an optimal alignment.\n",
    "    i = nx\n",
    "    j = ny\n",
    "    rx = []\n",
    "    ry = []\n",
    "    while i > 0 or j > 0:\n",
    "        if P[i,j] in [2, 5, 6, 9]:\n",
    "            rx.append(x[i-1])\n",
    "            ry.append(y[j-1])\n",
    "            i -= 1\n",
    "            j -= 1\n",
    "        elif P[i,j] in [3, 5, 7, 9]:\n",
    "            rx.append(x[i-1])\n",
    "            ry.append('-')\n",
    "            i -= 1\n",
    "        elif P[i,j] in [4, 6, 7, 9]:\n",
    "            rx.append('-')\n",
    "            ry.append(y[j-1])\n",
    "            j -= 1\n",
    "    # Reverse the strings.\n",
    "    rx = ''.join(rx)[::-1]\n",
    "    ry = ''.join(ry)[::-1]\n",
    "    return '\\n'.join([rx, ry, str(F[nx,ny])])\n",
    "    \n",
    "x = \"GATTACA\"\n",
    "y = \"GCATGCU\"\n",
    "print(nw(x, y))\n",
    "# G-ATTACA\n",
    "# GCA-TGCU\n",
    "\n",
    "np.random.seed(42)\n",
    "x = np.random.choice(['A', 'T', 'G', 'C'], 50)\n",
    "y = np.random.choice(['A', 'T', 'G', 'C'], 50)\n",
    "\n",
    "print(nw(x, y, gap = 0))\n",
    "# ----G-C--AGGCAAGTGGGGCACCCGTATCCT-T-T-C-C-AACTTACAAGGGT-C-CC-----CGT-T\n",
    "# GTGCGCCAGAGG-AAGT----CA--C-T-T--TATATCCGCG--C--AC---GGTACTCCTTTTTC-TA-\n",
    "\n",
    "print(nw(x, y, gap = 1))\n",
    "# GCAG-GCAAGTGG--GGCAC-CCGTATCCTTTC-CAAC-TTACAAGGGTCC-CCGT-T-\n",
    "# G-TGCGCCAGAGGAAGTCACTTTATATCC--GCGC-ACGGTAC-----TCCTTTTTCTA\n",
    "\n",
    "print(nw(x, y, gap = 2))\n",
    "# GCAGGCAAGTGG--GGCAC-CCGTATCCTTTCCAACTTACAAGGGTCCCCGTT\n",
    "# GTGCGCCAGAGGAAGTCACTTTATATCC-GCGCACGGTAC-TCCTTTTTC-TA\n",
    "# GCAG-GCAAGTGG--GGCAC-CCGTATCCTTTC-CAAC-TTACAAGGGTCC-CCGT-T-\n",
    "# G-TGCGCCAGAGGAAGTCACTTTATATCC--GCGC-ACGGTAC-----TCCTTTTTCTA\n",
    "\n",
    "print(nw(x, y, gap = 2))\n",
    "# GCAGGCAAGTGG--GGCAC-CCGTATCCTTTCCAACTTACAAGGGTCCCCGTT\n",
    "# GTGCGCCAGAGGAAGTCACTTTATATCC-GCGCACGGTAC-TCCTTTTTC-TA"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### FASTA Pseudocode\n",
    "\n",
    "### FASTA Pseudocode example\n",
    "\n",
    "'''\n",
    "fastaComparison(Q,S):  \n",
    "    q=len(Q)  \n",
    "    s=len(S)  \n",
    "    k=2 # choice of k  \n",
    "    #  \n",
    "    S=[0 for i in range(1-q,s-1)] # creation of S  \n",
    "    kmersQ = indexKmers(Q, k)  \n",
    "    kmersS = indexKmers(S, k)  \n",
    "    S = []  \n",
    "    for i in kmersQ:  \n",
    "        m = kmersQ[i]  \n",
    "        n = kmersQ[i]  \n",
    "        if (len(m) == 0):  \n",
    "            m = 0  \n",
    "        if (len(n) == 0):  \n",
    "            n = 0  \n",
    "        for mmer in m:  \n",
    "            for nmer in n:  \n",
    "                p = m[mmer] - n[nmer]  \n",
    "                S[p] += 1   \n",
    "    return S  \n",
    "'''  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### FASTA Implementation\n",
    "\n",
    "def indexKmers(S, k):\n",
    "    kmers = {}\n",
    "    lenseq = len(S)\n",
    "    for i in range(lenseq - k + 1):\n",
    "        substring = S[i:i+k]\n",
    "        if substring in kmers:\n",
    "            kmers[substring].append(i)\n",
    "        else:\n",
    "            kmers[substring]=[i]\n",
    "    \n",
    "    return(kmers)\n",
    "\n",
    "mykmers = list(indexKmers(\"ATTAGACACACAGTTTT\", 2))\n",
    "\n",
    "print(mykmers[1])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def fastaComparison(Q, S, k, lenlimit, gaplimit):\n",
    "    import re\n",
    "    # Q will be the query, S will be the target\n",
    "    q=len(Q)\n",
    "    s=len(S)    \n",
    "    #k=2 # choice of k\n",
    "\n",
    "    # q will be the 'negative index', s will be the 'positive index'   \n",
    "    F=[0 for i in range(1-q,s-1)] # creation of Diagonal Scores List\n",
    "    M=[\"\" for i in range(1-q,s-1)] # creation of Matches list\n",
    "    # Running kmers indexing for Q, S\n",
    "    kmersQ = indexKmers(Q, k)  \n",
    "    kmersS = indexKmers(S, k)\n",
    "\n",
    "    for i in kmersQ:  \n",
    "        m = kmersQ[i]  \n",
    "        n = kmersS[i]\n",
    "        if (len(m)!=0 and len(n)!=0):\n",
    "            for j in m:\n",
    "                for l in n: \n",
    "                    p = l - j \n",
    "                    F[p] += 1\n",
    "    \n",
    "   \n",
    "    matchedquery = []\n",
    "    matchedsequence = []\n",
    "\n",
    "    for d in range(1-q,s-1):\n",
    "        if F[d] == (q - k + 1):\n",
    "            matchedquery = Q\n",
    "            positions = [match.start() for match in re.finditer(matchedquery, S)] \n",
    "            matchedsequence = S[positions[0]:positions[0]+q]\n",
    "            print(\"Full match found at S in position\",\"=\",positions,\"\\n\",matchedquery,\"\\n\",matchedsequence)\n",
    "        if F[d] >= lenlimit:\n",
    "            print(d, F[d])\n",
    "            matchedquery = Q\n",
    "            positions = [match.start() for match in re.finditer(matchedquery, S)] \n",
    "            matchedsequence = S[positions[0]:positions[0]+F[d]+1]\n",
    "            print(\"Partial match found at S in position\",\"=\",positions,\"\\n\",matchedquery,\"\\n\",matchedsequence)\n",
    "    return(F)  \n",
    "\n",
    "print(fastaComparison('TCATTTTCCCACATTTT','ATCAGGCACATCATTTT', 2, 3, 2))\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Appendix\n",
    "#### Bits of code that can be useful"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the following line searches a substring on a longer string and records ALL positions where it is found\n",
    "# and stores them in a list\n",
    "import re\n",
    "import regex\n",
    "\n",
    "import time\n",
    "start_time = time.time()\n",
    "\n",
    "keystring = 'ATGCACGACC'\n",
    "dictkmers[keystring] = [m.start() for m in re.finditer(f'(?={keystring})', ecoli)]\n",
    "\n",
    "print(keystring, dictkmers[keystring])\n",
    "print(\"--- %s seconds ---\" % (time.time() - start_time))\n",
    "\n",
    "# Sorting a dictionary by key value\n",
    "#sorted_dictkmers = {key: value for key, value in sorted(dictkmers.items())}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the following calculates the Collatz Series and plots the intermediate steps until convergence for a given number N\n",
    "\n",
    "def collatz(n):\n",
    "    sequence = [n]\n",
    "    s = 0\n",
    "    while(n != 1):\n",
    "        if (n % 2 == 0):\n",
    "            n = n // 2\n",
    "        else:\n",
    "            n = 3*n + 1\n",
    "        sequence.append(n)\n",
    "        s += 1\n",
    "    #return \", \".join([str(x) for x in sequence])\n",
    "    return sequence\n",
    "\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import math \n",
    "\n",
    "n = 1000023\n",
    "y = [math.log10(x) for x in collatz(n)]\n",
    "x = list(range(1, len(y) + 1))\n",
    "\n",
    "plt.scatter(x, y, color='red', label='Data Points')\n",
    "\n",
    "# Connect the points with lines\n",
    "plt.plot(x, y, color='blue', linestyle='-', linewidth=2, label='Connecting Lines')\n",
    "\n",
    "# Set labels and title\n",
    "plt.xlabel('Step')\n",
    "plt.ylabel('Collatz Number (log10)')\n",
    "plt.title('Collatz Series of N')\n",
    "\n",
    "# Show legend\n",
    "plt.legend()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01J7KATZD75G13NPFZ3R05TPEN",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The following applies the Collatz function on a series of \n",
    "# numbers i from 1 up to N and then plots the number of steps S against i\n",
    "\n",
    "def collatz(n):\n",
    "    sequence = [n]\n",
    "    s = 0\n",
    "    while(n != 1):\n",
    "        if (n % 2 == 0):\n",
    "            n = n // 2\n",
    "        else:\n",
    "            n = 3*n + 1\n",
    "        sequence.append(n)\n",
    "        s += 1\n",
    "    #return \", \".join([str(x) for x in sequence])\n",
    "    return s\n",
    "\n",
    "def scollatz(n):\n",
    "    S = {}\n",
    "    for i in range(1, n+1):\n",
    "        S[i] = collatz(i)\n",
    "    return list(S.values())\n",
    "\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import math \n",
    "\n",
    "n = 10000\n",
    "y = scollatz(n)\n",
    "x = list(range(1, n+1))\n",
    "\n",
    "plt.scatter(x, y, color='red', label='Data Points')\n",
    "\n",
    "# Connect the points with lines\n",
    "#plt.plot(x, y, color='blue', linestyle='-', linewidth=2, label='Connecting Lines')\n",
    "\n",
    "# Set labels and title\n",
    "plt.xlabel('i')\n",
    "plt.ylabel('Collatz Steps for i')\n",
    "plt.title('Collatz Series Length')\n",
    "\n",
    "# Show legend\n",
    "plt.legend()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "916dbcbb3f70747c44a77c7bcd40155683ae19c65e1c03b4aa3499c5328201f1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
