# BC205
## Algorithms in Bioinformatics - MSc in Bioinformatics - University of Crete

This is the git repository on the Course on Algorithms in Bioinformatics (BC205) which forms part of the Program of the Masters in Bionformatics at the University of Crete

**Contact Details**  
*Christoforos Nikolaou: nikolaou@uoc.gr*  
*web: https://computational-genomics-uoc.weebly.com/*  
*twitter: @guilderstern*  

## 2018-2019 Academic Year
### We will be covering the following topics  

**[1. Introduction to Algorithms](https://github.com/christoforos-nikolaou/BC205/blob/master/BC205_Introduction_beamer.pdf): Types of Algorithms, Complexity and Iteration vs Recursion**  
  a. Finding the greatest common divisor of two integers  
  b. Calculating the Fibonacci series  
    
**[2. Algorithmic Techniques](https://github.com/christoforos-nikolaou/BC205/blob/master/BC205_Introduction_beamer.pdf): Brute Force vs Divide and Conquer, Binary Searches**  
  a. Sorting with Bubble Sort vs MergeSort  
  b. Solving a polynomial equation with Binary Search vs Newton-Raphson
    
**[3. Sequence Analysis](https://github.com/christoforos-nikolaou/BC205/blob/master/BC205_SeqAnalysis_beamer.pdf): Sequence Composition and the GC%, Counting k-mers in a sequence, Estimating o/e ratios of k-mers**   
  a. Finding the GC% of a sequence   
  b. Detecting regions in a sequence with extreme composition (based on a Z-score calculation)  
  c. Comparing sequences on the basis of their o/e k-mer ratios  
  
  **[Exercise 1](https://github.com/christoforos-nikolaou/BC205/blob/master/Exercise1_2019.md)**: Outlier Detection in Sequences/Sequence Comparisons with Genomic Signatures  
    
**[4. Motifs in Sequences I](https://github.com/christoforos-nikolaou/BC205/blob/master/cb_2016_lecture_03_motifs.pdf): Motif definition, Motif representation in PWM and PSSM, Motif Strength as Information Content**   
  a. Calculation of PWM and PSSM matrices from a set of motif instances   
  b. Calculation of motif strength as mean information content (Difference in Shannon Entropy)
  c. Locating motif instances in an unknown sequence based on a PWM or a PSSM
    
**[5. Motifs in Sequences II](https://github.com/christoforos-nikolaou/BC205/blob/master/BC205_MotifDiscovery_beamer.pdf): Motif Deduction, Greedy vs Randomized Approaches**   
  a. Estimating a motif from a set of sequences   
  b. Deduction of motifs through a randomized approach (Gibbs Sampling)  
  
  **Exercise 2**: Deduction of Sequence Motifs with Gibbs Sampling/Location of instances in an unknown sequence
  
**[6. NGS Algorithms](https://github.com/christoforos-nikolaou/BC205/blob/master/BC205_NGSMapping_beamer.pdf): Large-scale alignment and Genome-wide read mapping**   
  a. Heuristic approaches (BLAST, BLAT)   
  b. Alternative data structure approaches: Suffix Trees   
  c. Transformation approaches: Burrows-Wheeler Transform (BWT)  
  
  **Exercise 3**: BWT encoding and decoding of a given sequence
   
 **[7. Network Algorithms](https://github.com/christoforos-nikolaou/BC205/blob/master/BC205_GenomeScaleAlgorithms_beamer.pdf): Network Basics, Statistical Properties of Networks, Network Searches and Applications in NGS**   
  a. Estimating the scaling properties of a network through node-degree distributions   
  b. BFS and DFS for network analysis  
  c. Genome assembly and De Brujn Graphs   
