{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercises - Indicative Solutions - Discussion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Some auxiliary functions\n",
    "\n",
    "# Reading a fast file\n",
    "def readfasta(fastafile):\n",
    "    import regex as re\n",
    "    f=open(fastafile, 'r')\n",
    "    seq = \"\"\n",
    "    total = 0\n",
    "    for line in f:\n",
    "        x=re.match(\">\", line)\n",
    "        if x == None:\n",
    "            length=len(line)\n",
    "            total=total+length\n",
    "            seq=seq+line[0:length-1]\n",
    "    f.close()\n",
    "    return(seq)\n",
    "\n",
    "# A maximum function\n",
    "def maximum(x, y) :\n",
    "    if x > y :\n",
    "        return x\n",
    "    else :\n",
    "        return y\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Palindromes. Finding them, extracting the greatest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GVVVG'"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Take 1: Iteration \n",
    "\n",
    "def longestpalindrome(s):\n",
    "  palindromes=[]\n",
    "  for i in range(1,len(s)):\n",
    "    try:\n",
    "        m=1 \n",
    "        while s[i-m]==s[i+m]:\n",
    "            palindromes.append(s[i-m:i+m+1])\n",
    "            m+=1\n",
    "    except:\n",
    "        pass\n",
    "  return max(palindromes, key=len)\n",
    "\n",
    "# Run\n",
    "myseq=readfasta(\"../files/ecoli.fa\")\n",
    "myseq=ecoliseq.replace('N','')\n",
    "myseq='AATGVGVVVGCCCGXATAAAATAXX'\n",
    "longestpalindrome(myseq)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The solution above only works for odd-sized palindromes. \n",
    "What about even-sized ones?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A\n",
      "T\n",
      "G\n",
      "T\n",
      "V\n",
      "V\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'A'"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Take 2: Recursion\n",
    "\n",
    "# we will build on a simple recursive function like the one given below\n",
    "\n",
    "def palindrome(mystring):\n",
    "    if len(mystring) < 2:\n",
    "        return True\n",
    "    if mystring[0] != mystring[-1]:\n",
    "        return False\n",
    "    return palindrome(mystring[1:-1])\n",
    "\n",
    "\n",
    "# We will next scan the sequence of the genome and call the function \n",
    "\n",
    "def recursivePalindrome(s):\n",
    "    palindromes = []\n",
    "    for i in range(1,len(s)):\n",
    "        k = 0\n",
    "        subsequence = s[i]\n",
    "        if (palindrome(subsequence)):\n",
    "            print(subsequence)\n",
    "            palindromes.append(subsequence)\n",
    "            k += 1\n",
    "            subsequence = s[i-k:i+k]\n",
    "        else:\n",
    "            next            \n",
    "            \n",
    "    return max(palindromes, key=len)\n",
    "\n",
    "recursivePalindrome(\"AATGTVV\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Non-mers in E.coli Genome\n",
    "\n",
    "Below I provide code taken from one of your assignments, slightly modified and with comments."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found: 223321  in the genome of E.coli\n",
      "The number of non 10-mers is: 223321\n",
      "Some of those non-mers are ['TCACTTACCA', 'ACAGCTGTTA', 'ATGAGGCTAG', 'TTTCGTGATT', 'GCCTCGGGCG', 'ACTAACTTCC', 'CATTCTATTA', 'TCGATTCAAG', 'ACATATTAAG', 'GCCTATGGTA']\n",
      "According to theory we would expect 223302 non-mers since all the 10-mers are 1048576 and the observed in E.Coli are 825274\n",
      "223321\n",
      "['TAGGGTCACG', 'GTACACGAGG', 'CACCGTATAG', 'TCGTTTAGAG', 'TTTAGTCGCG', 'GATCTCATGT', 'AAGGACTGTT', 'GCCCCACCGA', 'CATCTCACAC', 'AGAATGCTAA']\n"
     ]
    }
   ],
   "source": [
    "with open (\"../files/ecoli.fa\") as l:\n",
    "    l.readline()\n",
    "    ecolseq =\"\"\n",
    "    for ln in l:\n",
    "        lnseq = ln.strip()\n",
    "        ecolseq += lnseq\n",
    "\n",
    "#Findind all the possible 10-mers that can be found with the 4 nucleotides\n",
    "import itertools\n",
    "\n",
    "k=10\n",
    "bases = [\"A\",\"T\",\"G\",\"C\"]\n",
    "combos = itertools.product(bases, repeat=k)\n",
    "strcomb = (\"\".join(y) for y in combos) #generators save memory compared to lists\n",
    "\n",
    "#Find all k-mers/10-mers in the genome\n",
    "\n",
    "kmers=[]\n",
    "for i in range(len(ecolseq)-k+1):\n",
    "    kmers.append(ecolseq[i:i+k])\n",
    "\n",
    "#Isolation of all the unique ones (makes the procedure faster)\n",
    "kmers = set(kmers)\n",
    "\n",
    "#Calculating how many 10-mers are not included in the E.coli genome\n",
    "count=0\n",
    "nonkmers =[]\n",
    "allk =0\n",
    "res =0\n",
    "for y in strcomb:\n",
    "    allk += 1     #control to check if the theoritical 4^10 has been produced with the product functional tool \n",
    "    if y not in kmers:\n",
    "        count +=1 #counting the 10-mers not included is same as len(nonkemers)/not necessary / check point\n",
    "        nonkmers.append(y)\n",
    "    else:         #not necessary / check point\n",
    "        res+=1    #counting the included 10-mers / check point\n",
    "\n",
    "\n",
    "import random\n",
    "\n",
    "print(\"The number of non 10-mers is:\",count)\n",
    "print(\"Some of those non-mers are\", random.sample(nonkmers,10))\n",
    "\n",
    "# POINT: We can do the same without looping, simply by comparing sets\n",
    "allkmers = set([''.join(p) for p in itertools.product(bases, repeat=k)])\n",
    "nonmers = allkmers.difference(kmers)\n",
    "print(\"Found:\", len(nonmers), \"nonmers in the genome of E.coli\")\n",
    "\n",
    "#Expected non-10-mers/ Theoritical check: \n",
    "theor = 4**10\n",
    "obs = len(kmers)\n",
    "exp = theor - obs\n",
    "print(\"According to theory we would expect\", exp,\"non-mers since all the 10-mers are\", theor,\"and the observed in E.Coli are\", obs)\n",
    "\n",
    "# NOTE: The theoritical check is not in agreement with the ones found above with comparison. Not quite sure why.\n",
    "\n",
    "# =================================================================================================================\n",
    "# POINT: This is not correct from a theoretical viewpoint. In fact it is much more complex. Can you think of why?\n",
    "# =================================================================================================================\n",
    "\n",
    "\n",
    "#Another way to avoid for-loops and exhausting search is to immediately produce sets that contain unique 10-mers and calculate their difference:\n",
    "#The reading of the file is common\n",
    "\n",
    "# -> This is the point made above! (OK)\n",
    "\n",
    "kmers2 = {ecolseq[i:i+k] for i in range(len(ecolseq)-k+1) }\n",
    "combos = itertools.product(bases, repeat=k) \n",
    "#must to re-write the command since it's a generator that has been used above and needs to be generated to reused\n",
    "strcomb2 = {\"\".join(y) for y in combos} #permuation with replacement we can make it a set as they all uniq\n",
    "nonmers = strcomb2 - kmers2\n",
    "\n",
    "print(len(nonmers))\n",
    "print(random.sample(nonmers, 10)) #demonstartion of some non-mers"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Question: Did anyone try to use the binary search? Would it be faster?"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Gibbs Sampler\n",
    "\n",
    "Below I am attaching an indicative solution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The information content of the motif divided by it's length is: 2.0\n",
      "The pwm of the motif is:\n",
      " [[0. 0. 0.]\n",
      " [0. 0. 1.]\n",
      " [1. 0. 0.]\n",
      " [0. 1. 0.]]\n",
      "The motif is: CGT\n",
      "\n",
      "The information content of the motif divided by it's length is: 1.8640950668139766\n",
      "The pwm of the motif is:\n",
      " [[0.   1.   0.   0.92]\n",
      " [0.02 0.   1.   0.08]\n",
      " [0.   0.   0.   0.  ]\n",
      " [0.98 0.   0.   0.  ]]\n",
      "The motif is: GATA\n",
      "\n",
      "The information content of the motif divided by it's length is: 1.5973762912220875\n",
      "The pwm of the motif is:\n",
      " [[0.6  0.   1.   0.   0.94]\n",
      " [0.24 0.   0.   0.98 0.  ]\n",
      " [0.12 0.   0.   0.   0.02]\n",
      " [0.04 1.   0.   0.02 0.04]]\n",
      "The motif is: AGATA\n",
      "\n",
      "The information content of the motif divided by it's length is: 1.3910716042455649\n",
      "The pwm of the motif is:\n",
      " [[0.48 0.   1.   0.   0.88 0.44]\n",
      " [0.46 0.02 0.   1.   0.12 0.16]\n",
      " [0.02 0.   0.   0.   0.   0.04]\n",
      " [0.04 0.98 0.   0.   0.   0.36]]\n",
      "The motif is: AGATAA\n",
      "\n",
      "The information content of the motif divided by it's length is: 1.1870405050107355\n",
      "The pwm of the motif is:\n",
      " [[0.66 0.   1.   0.   0.82 0.48 0.2 ]\n",
      " [0.28 0.04 0.   1.   0.1  0.1  0.4 ]\n",
      " [0.06 0.   0.   0.   0.04 0.02 0.12]\n",
      " [0.   0.96 0.   0.   0.04 0.4  0.28]]\n",
      "The motif is: AGATAAT\n"
     ]
    }
   ],
   "source": [
    "# Gibbs Sampling to locate a motif in a set of sequences\n",
    "# code\n",
    "sequences = []\n",
    "with open ('../files/motifs_in_sequence.fa') as file:\n",
    "    for line in file:\n",
    "        sequences.append(line.strip()) # 50 sequences as elements of a list. 100 bases each sequence\n",
    "    \n",
    "\n",
    "##Gibbs sampler##\n",
    "import random\n",
    "import numpy as np\n",
    "\n",
    "def Gibbs_sampler(sequences,k): #k is the length of the motif, sequence is a list with the sequences\n",
    "    \n",
    "    dictionary = {'A':0,\n",
    "                  'T':1,\n",
    "                  'C':2,\n",
    "                  'G':3}\n",
    "    \n",
    "    column_sum = len(sequences) #number of rows (50) or number of sequences\n",
    "    length = len(sequences[0]) #number of columns or number of nucleotides in seq\n",
    "    Imax = 1.8*k #threshold of I\n",
    "    \n",
    "    pwm = np.zeros([4,k]) # A,T,C,G X len(motif)\n",
    "    \n",
    "    for seq in sequences:\n",
    "        rand_start = random.randint(0, length-k) #pick a random nucleotide from each sequence\n",
    "        motif = seq[rand_start:rand_start+k] #and take substring as the motif\n",
    "       \n",
    "        lst = enumerate(motif) #finding the index of each nucleotide in the motif to access the correct column\n",
    "                               #and using the dictionary to access the correct row \n",
    "        for i in lst: #making the first random pwm\n",
    "            pwm[dictionary[i[1]],i[0]]+=1\n",
    "            \n",
    "    pwm = pwm/column_sum\n",
    "    \n",
    "    information = np.zeros([1,k])\n",
    "    count=0\n",
    "    while (np.sum(information)) < Imax: #while information_content of the pwm \n",
    "        motives=[]                      #is less than the threshold\n",
    "        \n",
    "        information_old = np.sum(information) #keeping the previous value of information contect\n",
    "                                             #to check convergence in case the theshold \n",
    "        for row in range (column_sum):                #is never reached\n",
    "            maxx=0\n",
    "            rand_seq = random.randint(0, column_sum-1) #pick a random index - sequence\n",
    "            seq = sequences[rand_seq]               \n",
    "            for i in range(len(seq)-k):   #take each k-mer from the sequence \n",
    "                score = 0\n",
    "                substring = seq[i:i+k]\n",
    "                lst = enumerate(substring)\n",
    "                \n",
    "                for j in lst:                         #scoring each k-mer based on the pwm\n",
    "                    score+=pwm[dictionary[j[1]],j[0]]   #keeping the motif with the highest score\n",
    "                                                         #from each sequence\n",
    "                if score > maxx:  \n",
    "                    maxx = score\n",
    "                    motif = substring\n",
    "                    \n",
    "            motives += [motif] #keep all the motifs with the highest score in the list motives\n",
    "        \n",
    "        pwm = np.zeros([4,k]) # A,T,C,G X len(motif) \n",
    "        \n",
    "        for elem in motives: \n",
    "            lst = enumerate(elem)\n",
    "            for i in lst:         #making the new pwm\n",
    "                pwm[dictionary[i[1]],i[0]]+=1\n",
    "                 \n",
    "        pwm = pwm/column_sum\n",
    "        \n",
    "        information = np.zeros([1,k]) #computing the information of each position\n",
    "        for i in range(k):\n",
    "            information[0,i] = 2-abs(sum([elem*np.log2(elem) for elem in pwm[:,i] if elem > 0]))\n",
    "            \n",
    "\n",
    "        if abs(information_old - np.sum(information)) <= 0.5: #ckecking convergence \n",
    "            count+=1\n",
    "            if count == 10: #if the difference of the information content is less or equal to 0.5\n",
    "                break       #for consecutive 10 iterations then break\n",
    "        else:\n",
    "            count=0\n",
    "    \n",
    "    max_index_col = np.argmax(pwm, axis=0) #extracting the motif according to the   \n",
    "                                           #highest frequency of each nucleotide in each position\n",
    "    motif=''\n",
    "    for values in max_index_col:\n",
    "        for keys in dictionary.keys():\n",
    "            if values == dictionary[keys]:\n",
    "                motif+= keys\n",
    "        \n",
    "    return pwm,information,motif\n",
    "\n",
    "\n",
    "#repeat the algorithmm 100 times for each k (3 to 7) and keep the pwm and motif with the highest infromation_content\n",
    "#this process takes approximately 4min (in my computer)\n",
    "    \n",
    "####100-cycled GIbbs#####\n",
    "for k in range (3,8):\n",
    "    highest_info = 0\n",
    "    for i in range (100):\n",
    "        summ=0\n",
    "        pwm, information_content,motif = Gibbs_sampler(sequences,k)\n",
    "        summ+=np.sum(information_content)\n",
    "        if summ > highest_info:\n",
    "            highest_info = summ\n",
    "            pwm_ret = pwm\n",
    "            motif_ret = motif\n",
    "        \n",
    "    print('\\nThe information content of the motif divided by it\\'s length is:',highest_info/k) #divide by length to normalize and compare among other k\n",
    "    print('The pwm of the motif is:\\n',pwm_ret)\n",
    "    print('The motif is:',motif_ret)\n",
    "\n",
    "\n",
    "\n",
    "#To find the motifs for each k with the highest information content i have to repeat gibbs sampler many times because of the randomness that takes place\n",
    "#The motif that returns the highest scaled information content is GAT (k = 3, I/k = 1.9528 or 2!) but i think that the motif that we are looking for\n",
    "#is the motif GATA (k=4, I/k = 1.857) which contains GAT and is contained in all the longer found motifs.\n",
    "#(Also we know the existence of the GATA transcription factors and indeed GATA is part of these binding sites)\n",
    "#Sometimes the algorithm returns other 3-mers (eg ATG,GGC,AAG) with high information content as well but it\n",
    "#is noted that GATA is returned almost in every repetition of the 100-cycled Gibbs which means that it may be the only 4-mer motif with so high information content.\n",
    "#Finally i have to report that the threshold that i chose for checking convergence may not be the best choice and should be stricter\n",
    "#Either way i kept this threshold because otherwise i would have made the algorithm a lot slower (is already slow though :P)."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
