{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pwm(lista):\n",
    "    nuc = ['A', 'C', 'G', 'T']\n",
    "    profile=[[0 for i in range(len(lista[0]))] for j in range(len(nuc))]\n",
    "    for instance in lista:\n",
    "        for j in range(len(instance)):\n",
    "            residue=instance[j]\n",
    "            profile[nuc.index(residue)][j]+=1\n",
    "            profile[nuc.index(residue)][j]=float(profile[nuc.index(residue)][j])\n",
    "    import numpy as np\n",
    "    pwm = np.array(profile)\n",
    "    pwm = pwm/len(lista)\n",
    "    return(pwm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "import regex as re\n",
    "\n",
    "def kmers(genomefile, k):\n",
    "    kmertable = {} \n",
    "    for i in range(len(seq)-k):\n",
    "        DNA=seq[i:i+k]\n",
    "        if DNA not in kmertable.keys():\n",
    "            kmertable[DNA]=1\n",
    "        else:\n",
    "            kmertable[DNA]+=1\n",
    "\n",
    "    kmertable = {k: float(v) / len(seq) for k, v in kmertable.items()}\n",
    "    kmertable = dict(sorted(kmertable.items()))\n",
    "    return(list(kmertable.values()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pssm(pwm, nucfreqs):\n",
    "    import numpy as np\n",
    "    import math\n",
    "    pseudocount=0.01\n",
    "    pssm=[[0 for i in range(len(pwm[0]))] for j in range(len(nucfreqs))]\n",
    "    for i in range(len(nucfreqs)):\n",
    "        pssm[i]=(np.array(pwm[i])+pseudocount)/nucfreqs[i]\n",
    "    for i in range(len(pssm)):\n",
    "        for k in range(len(pssm[0])):\n",
    "            pssm[i][k]=math.log(pssm[i][k])/math.log(2)\n",
    "    return(np.array(pssm))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pssmSearch(pssm, sequence, threshold):\n",
    "    nuc = ['A', 'C', 'G', 'T']\n",
    "    hits = []\n",
    "    instances = []\n",
    "    x = []\n",
    "    allscores = []\n",
    "    maxPssm = 0\n",
    "    for j in range(len(pssm[0])):\n",
    "        maxPssm = maxPssm + max(pssm[:,j])\n",
    "    \n",
    "    for i in range(len(sequence)-len(pssm[0])):\n",
    "        x.append(i)\n",
    "        instance=sequence[i:i+len(pssm[0])]\n",
    "        score=0\n",
    "        for l in range(len(instance)):\n",
    "            score=score+pssm[nuc.index(instance[l])][l]\n",
    "        if (score > threshold*maxPssm):\n",
    "            hits.append(i)\n",
    "            instances.append(instance)\n",
    "            allscores.append(score)\n",
    "    \n",
    "    return(instances)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pwmEntropyInformation(pwm):\n",
    "    import numpy as np\n",
    "    k = pwm.shape[1]\n",
    "\n",
    "    information = np.zeros([1,k])\n",
    "    for i in range(k):\n",
    "        information[0,i] = 2-abs(sum([elem*np.log2(elem) for elem in pwm[:,i] if elem > 0]))\n",
    "    \n",
    "    sumInfo = np.sum(information)\n",
    "    scaledSumInfo = sumInfo/k\n",
    "    \n",
    "    return(scaledSumInfo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.0\n",
      "k = 3\n",
      "CAA\n",
      "[[0.16 0.4  0.28]\n",
      " [0.34 0.18 0.24]\n",
      " [0.2  0.18 0.26]\n",
      " [0.3  0.24 0.22]] \n",
      "\n",
      "2.0\n",
      "k = 4\n",
      "TATC\n",
      "[[0.22 0.34 0.2  0.28]\n",
      " [0.26 0.2  0.1  0.3 ]\n",
      " [0.2  0.22 0.32 0.18]\n",
      " [0.32 0.24 0.38 0.24]] \n",
      "\n",
      "1.6235250802274563\n",
      "1.6143684940767085\n",
      "2.0\n",
      "k = 5\n",
      "TACGA\n",
      "[[0.14 0.34 0.28 0.2  0.34]\n",
      " [0.24 0.16 0.34 0.22 0.24]\n",
      " [0.26 0.28 0.24 0.34 0.3 ]\n",
      " [0.36 0.22 0.14 0.24 0.12]] \n",
      "\n",
      "1.6818553323583447\n",
      "1.8469506943242517\n",
      "k = 6\n",
      "GTGTCT\n",
      "[[0.24 0.28 0.26 0.2  0.28 0.12]\n",
      " [0.2  0.18 0.14 0.28 0.38 0.3 ]\n",
      " [0.36 0.22 0.36 0.18 0.12 0.26]\n",
      " [0.2  0.32 0.24 0.34 0.22 0.32]] \n",
      "\n",
      "1.8571428571428572\n",
      "k = 7\n",
      "ATCGTTG\n",
      "[[0.44 0.24 0.16 0.22 0.26 0.18 0.22]\n",
      " [0.2  0.2  0.28 0.14 0.24 0.22 0.28]\n",
      " [0.26 0.26 0.28 0.32 0.18 0.26 0.28]\n",
      " [0.1  0.3  0.28 0.32 0.32 0.34 0.22]] \n",
      "\n"
     ]
    }
   ],
   "source": [
    "import regex as re\n",
    "import random\n",
    "\n",
    "for k in range(3,8):\n",
    "    fn = \"motifs_in_sequence.fa\"\n",
    "    \n",
    "    #Make a list with all the 50 sequences\n",
    "    seq_list = []\n",
    "    with open (fn, \"r\") as file:\n",
    "        for line in file:\n",
    "            seq_list.append(line)\n",
    "    \n",
    "    #Make a string that contains all the sequences\n",
    "    seq = \"\"\n",
    "    for string in seq_list:\n",
    "        seq+=string[:-1] #until -1 to ignore the '\\n'\n",
    "    \n",
    "    #Find the frequences of each nucleotide in whole concatenated sequence\n",
    "    nucfreqs = kmers(fn,1)\n",
    "    \n",
    "    I=0\n",
    "    while I < 1.8:\n",
    "        #Make a list with random kmers of length k\n",
    "        random_kmers=[]\n",
    "        for i in seq_list[:50]:\n",
    "            seqs_kmers=[]\n",
    "            for j in range(len(i)-k):\n",
    "                seqs_kmers.append((i[j:j+k]))\n",
    "            random_kmers.append(random.choice(seqs_kmers))\n",
    "\n",
    "        mypwm=pwm(random_kmers)\n",
    "        mypssm=pssm(mypwm, nucfreqs)\n",
    "        pssmsearches = pssmSearch(mypssm, seq, 0.95)\n",
    "        I=pwmEntropyInformation(pwm(pssmsearches))\n",
    "        print(I)\n",
    "\n",
    "        if I >= 1.8:\n",
    "            print (f\"k =\",k)\n",
    "            print(pssmsearches[0])\n",
    "            print(mypwm, '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
